ORA RNA-seq: Part 1 - The Set Up (#394)
AlexsLemonade/refinebio-examples: 03-rnaseq/pathway-analysis_rnaseq_01_ora.Rmd

---
title: "Over-representation analysis - RNA-Seq"
author: "CCDL for ALSF"

> date: "`r format(Sys.time(), '%B %Y')`"

Reviewer: Set date manually?
Appears this was missed in Microarray ORA too.

We highly recommend taking a look at [Ten Years of Pathway Analysis: Current Approaches and Outstanding Challenges](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375) from @Khatri2012 for a more comprehensive overview. We have provided primary publications and documentation of the methods we will introduce below as well as some recommended reading in the [`Resources for further learning` section](#resources-for-further-learning).

### How to choose a pathway analysis?

This table summarizes the pathway analyses examples in this module.

|Analysis|What is required for input|What output looks like |✅ Pros| ⚠️ Cons|
|--------|--------------------------|-----------------------|-------|-------|
|[**ORA (Over-representation Analysis)**](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/pathway-analysis_rnaseq_01_ora.html)|A list of gene IDs (no stats needed)|A per-pathway hypergeometric test result|- Simple<br><br> - Inexpensive computationally to calculate p-values| - Requires arbitrary thresholds and ignores any statistics associated with a gene<br><br> - Assumes independence of genes and pathways|

> |[**GSEA (Gene Set Enrichment Analysis)**](https://alexslemonade.github.io/refinebio-examples/03-rnaseq/pathway-analysis_rnaseq_02_gsea.html)|A list of genes IDs with gene-level summary statistics|A per-pathway enrichment score|- Includes all genes (no arbitrary threshold!)<br><br> - Attempts to measure coordination of genes|- Permutations can be expensive<br><br> - Does not account for pathway overlap<br><br> - Two-group comparisons not always appropriate/feasible|

Author: Note that I've included these links, but they are broken now. This means we will have to be a bit cognizant about making the other RNA-seq pathway analyses all live at the same time? Let me know if there's something else you suggest.

Reviewer: I can't think of a solution to this problem, except to not include these links until each part goes in, which seems messy...

if (!dir.exists("data")) {
  dir.create("data")
}

# Define the file path to the plots directory
plots_dir <- "plots" # Can replace with path to desired output plots directory

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the results directory
results_dir <- "results" # Can replace with path to desired output results directory

# Create the results folder if it doesn't exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

In the same place you put this `.Rmd` file, you should now have three new empty folders called `data`, `plots`, and `results`!

**REVIEW**

## About the dataset we are using for this example

For this example analysis, we will use this [acute viral bronchiolitis dataset](https://www.refine.bio/experiments/SRP140558).
The data that we downloaded from refine.bio for this analysis has 62 paired peripheral blood mononuclear cell RNA-seq samples obtained from 31 patients.
Samples were collected at two time points: during their first, acute bronchiolitis visit (abbreviated "AV") and their recovery, their post-convalescence visit (abbreviated "CV").

> We used this dataset, to identify co-expressed genes in an [example analysis](https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/network-analysis_rnaseq_01_wgcna.html) using [`WGCNA`](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/) [@Langfelder2008].

Reviewer: Minor wording change.
```suggestion
We used this dataset to identify modules of co-expressed genes in an [example analysis](https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/network-analysis_rnaseq_01_wgcna.html) using [`WGCNA`](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/) [@Langfelder2008].
```

> We have provided this file for you and the code in this notebook will read in the results that are stored online, but if you'd like to follow the steps for obtaining this results file yourself, we suggest going through [that WGCNA example](https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/network-analysis_rnaseq_01_wgcna.html).

Reviewer: Breaking up a long sentence here, with minor edits.

```suggestion
We have provided this file for you and the code in this notebook will read in the results that are stored online.
If you'd like to follow the steps for creating this results file from the refine.bio data, we suggest going through [that WGCNA example](https://alexslemonade.github.io/refinebio-examples/04-advanced-topics/network-analysis_rnaseq_01_wgcna.html).
```

In this example, we will focus on one gene module we identified through WGCNA, called module 19, to see what pathways these genes in this module may be associated with. 

> We identified this module 19 as the most differentially expressed between the datasets' two time points (during illness and recovering from illness).

Reviewer: ```suggestion
Module 19 was the most differentially expressed between the datasets' two time points (during illness and recovering from illness).
```

> Here's a heatmap to show a preview of this module and the genes that make it up. 

Reviewer: I'm rephrasing this for a bit more clarity.

```suggestion
The heatmap below summarizes the expression of the genes that make up module 19. 
```

<img src="https://refinebio-examples.s3.us-east-2.amazonaws.com/04-advanced-topics/results/SRP140558_module_19_heatmap.png" width = 500/>

Each row is a gene that is a member of module 19, and the composite expression of these genes, called an eigengene, is shown in the barplot below. 
This plot demonstrates how these genes, together as a module, are differentially expressed between the two time points. 

**IDENTICAL TO MICROARRAY EXAMPLE**
## Check out our file structure!

Your new analysis folder should contain:

- The example analysis `.Rmd` you downloaded  
- A folder called `data` (currently empty)
- A folder for `plots` (currently empty)  
- A folder for `results` (currently empty)  

Your example analysis folder should contain your `.Rmd` and three empty folders (which won't be empty for long!).

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds).

**REVIEW**
# Using a different refine.bio dataset with this analysis?

> If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend replacing the `gene_module_url` with a different file path to a gene list that you are interested in. 
> The file we use here has two columns from our WGCNA module results, so if your gene list differs, many steps will need to be changed or deleted entirely depending on what your gene list file looks like (particularly in the [`Determine our genes of interest list` section](#determined-our-genes-of-interest-list)).

Reviewer: Here I am trying to be a bit more explicit about what the input file contains, so people have a bit more of a sense of what the minimal data required is. It is a bit repetitive with later information, but I think it is useful to include here.

```suggestion
The file we use here has two columns from our WGCNA module results: the id of each gene and the module it is part of. 
If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend replacing the `gene_module_url` with a different file path to a read in a similar table of genes with the information that you are interested in. 
If your gene table differs, many steps will need to be changed or deleted entirely depending on the contents of that file (particularly in the [`Determine our genes of interest list` section](#determined-our-genes-of-interest-list)).
```

```{r}
if (!("clusterProfiler" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("clusterProfiler", update = FALSE)
}

# This is required to make one of the plots that clusterProfiler will make
if (!("ggupset" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("ggupset", update = FALSE)
}

if (!("msigdbr" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("msigdbr", update = FALSE)
}

if (!("org.Hs.eg.db" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("org.Hs.eg.db", update = FALSE)
}
```

Attach the packages we need for this analysis.

```{r}
# Attach the library
library(clusterProfiler)

# Package that contains MSigDB gene sets in tidy format
library(msigdbr)

# Homo sapiens annotation package we'll use for gene identifier conversion
library(org.Hs.eg.db)

# We will need this so we can use the pipe: %>%
library(magrittr)
```

**REVIEW** 
## Import data

> For ORA, we only need a list of genes as our input, so this example can work for any situations where you have gene list and want to know more about what biological pathways it shares genes with.

Reviewer: ```suggestion
For ORA, we only need a list of genes of interest and a background gene set as our input, so this example can work for any situations where you have gene list and want to know more about what biological pathways are significantly represented.
```

We will perform ORA on one of the modules we identified in the WGCNA analysis but the rest of the genes will be used as "background genes".

Instead of using this URL below, you can use a file path to a TSV file with your desired gene list.
First we will assign the URL to its own variable called, `gene_module_url`.

```{r}
# Define the url to your gene list file
gene_module_url <- "https://refinebio-examples.s3.us-east-2.amazonaws.com/04-advanced-topics/results/SRP140558_wgcna_gene_to_module.tsv"
```

Read in the file that has WGCNA gene and module results.
Here we are using the URL we set up above, but this can be a local file path instead _i.e._ you can replace `gene_module_url` in the code below with a path to file you have on your computer like: `file.path("results", "SRP140558_results_gene_list.tsv")`.

```{r}
# Read in the contents of your differential expression results file
# `gene_module_url` can be replaced with a file path to a TSV file with your
# desired gene list results
gene_module_df <- readr::read_tsv(gene_module_url)
```

`read_tsv()` can read TSV files online and doesn't necessarily require you download the file first.

> Let's take a look at what this gene module list looks like.

Reviewer: I didn't like the "look at... looks like" here. Also, I want to be clear about list vs. table, since they are different in R and I don't want people confused. (Don't @ me that a data frame is a special list)

```suggestion
Let's take a look at this gene module table.
```
